server/src/main/java/com/vaadin/server/VaadinService.java

```diff
@@ -1860,17 +1859,13 @@ public void runPendingAccessTasks(VaadinSession session) {
             return;
         }
 
-        Map<Class<?>, CurrentInstance> oldInstances = CurrentInstance
-                .getInstances(false);
-
         FutureAccess pendingAccess;
+        Map<Class<?>, CurrentInstance> oldInstances = CurrentInstance
+                .setCurrent(session);
         try {
             while ((pendingAccess = session.getPendingAccessQueue()
                     .poll()) != null) {
                 if (!pendingAccess.isCancelled()) {
-                    CurrentInstance.clearAll();
-                    CurrentInstance.restoreInstances(
-                            pendingAccess.getCurrentInstances());
                     CurrentInstance.setCurrent(session);
```

Comment:
This line was removed from the original change, but it seems to me it's better to have it here